# Video Game Analysis

## 1. Import and Clean Data

In this section, the data are imported, checked for missing values and duplicates, and a column for total sales is added. 

### 1a. Import Libraries and Data

In [55]:
# Import Libraries
import numpy as np
import pandas as pd
import plotly_express as px

In [56]:
# Read in dataset
url = "https://raw.githubusercontent.com/kellyshreeve/Integrated_Project_1/main/moved_games.csv"
vg = pd.read_csv(url)

In [57]:
# View first 15 rows of the dataset 
vg.head(15)

Name Platform  Year_of_Release         Genre  \
0                    Wii Sports      Wii           2006.0        Sports   
1             Super Mario Bros.      NES           1985.0      Platform   
2                Mario Kart Wii      Wii           2008.0        Racing   
3             Wii Sports Resort      Wii           2009.0        Sports   
4      Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                        Tetris       GB           1989.0        Puzzle   
6         New Super Mario Bros.       DS           2006.0      Platform   
7                      Wii Play      Wii           2006.0          Misc   
8     New Super Mario Bros. Wii      Wii           2009.0      Platform   
9                     Duck Hunt      NES           1984.0       Shooter   
10                   Nintendogs       DS           2005.0    Simulation   
11                Mario Kart DS       DS           2005.0        Racing   
12  Pokemon Gold/Pokemon Silver       GB           1999.0  Role-Playing   
13                      Wii Fit      Wii           2007.0        Sports   
14           Kinect Adventures!     X360           2010.0          Misc   

    NA_sales  EU_sales  JP_sales  Other_sales  Critic_Score User_Score Rating  
0      41.36     28.96      3.77         8.45          76.0          8      E  
1      29.08      3.58      6.81         0.77           NaN        NaN    NaN  
2      15.68     12.76      3.79         3.29          82.0        8.3      E  
3      15.61     10.93      3.28         2.95          80.0          8      E  
4      11.27      8.89     10.22         1.00           NaN        NaN    NaN  
5      23.20      2.26      4.22         0.58           NaN        NaN    NaN  
6      11.28      9.14      6.50         2.88          89.0        8.5      E  
7      13.96      9.18      2.93         2.84          58.0        6.6      E  
8      14.44      6.94      4.70         2.24          87.0        8.4      E  
9      26.93      0.63      0.28         0.47           NaN        NaN    NaN  
10      9.05     10.95      1.93         2.74           NaN        NaN    NaN  
11      9.71      7.47      4.13         1.90          91.0        8.6      E  
12      9.00      6.18      7.20         0.71           NaN        NaN    NaN  
13      8.92      8.03      3.60         2.15          80.0        7.7      E  
14     15.00      4.89      0.24         1.69          61.0        6.3      E

In [58]:
# Print dataset info
vg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16713 non-null  object 
 1   Platform         16715 non-null  object 
 2   Year_of_Release  16446 non-null  float64
 3   Genre            16713 non-null  object 
 4   NA_sales         16715 non-null  float64
 5   EU_sales         16715 non-null  float64
 6   JP_sales         16715 non-null  float64
 7   Other_sales      16715 non-null  float64
 8   Critic_Score     8137 non-null   float64
 9   User_Score       10014 non-null  object 
 10  Rating           9949 non-null   object 
dtypes: float64(6), object(5)
memory usage: 1.4+ MB


This dataset has a total of 16715 entries. There are missing values in Name, Year_of_Release, Genre, Critic_Score, User_Score, and Rating. Year_of_Release needs to be changed to int data type and and User_Score needs to be changed to float data type. Variable names should be changed to snake case.

### 1b. Rename Columns and Fix Data Types

In [59]:
# Change variable names to snake case
vg = vg.rename(
    columns={'Name':'name',
             'Platform':'platform',
             'Year_of_Release':'year_of_release',
             'Genre':'genre',
             'NA_sales':'na_sales',
             'EU_sales':'eu_sales',
             'JP_sales':'jp_sales',
             'Other_sales':'other_sales',
             'Critic_Score':'critic_score',
             'User_Score':'user_score',
             'Rating':'rating'    
})

print(vg.columns)

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')


In [60]:
# Change year_of_release to int data type

# Frist, check unique values of 'year_of_release'
print(vg['year_of_release'].unique())

[2006. 1985. 2008. 2009. 1996. 1989. 1984. 2005. 1999. 2007. 2010. 2013.
 2004. 1990. 1988. 2002. 2001. 2011. 1998. 2015. 2012. 2014. 1992. 1997.
 1993. 1994. 1982. 2016. 2003. 1986. 2000.   nan 1995. 1991. 1981. 1987.
 1980. 1983.]


The values are all whole numbers, so it is safe to convert 'year_of_release' to int.

In [61]:
# Convert 'year_of_release' to int
vg['year_of_release'] = vg['year_of_release'].astype('Int64')

type = vg['year_of_release'].dtypes
print(f'The dtype for "year_of_release" now is: {type}')

The dtype for "year_of_release" now is: Int64


In [62]:
# Print unique values in 'user_score'
print(vg['user_score'].unique())

['8' nan '8.3' '8.5' '6.6' '8.4' '8.6' '7.7' '6.3' '7.4' '8.2' '9' '7.9'
 '8.1' '8.7' '7.1' '3.4' '5.3' '4.8' '3.2' '8.9' '6.4' '7.8' '7.5' '2.6'
 '7.2' '9.2' '7' '7.3' '4.3' '7.6' '5.7' '5' '9.1' '6.5' 'tbd' '8.8' '6.9'
 '9.4' '6.8' '6.1' '6.7' '5.4' '4' '4.9' '4.5' '9.3' '6.2' '4.2' '6' '3.7'
 '4.1' '5.8' '5.6' '5.5' '4.4' '4.6' '5.9' '3.9' '3.1' '2.9' '5.2' '3.3'
 '4.7' '5.1' '3.5' '2.5' '1.9' '3' '2.7' '2.2' '2' '9.5' '2.1' '3.6' '2.8'
 '1.8' '3.8' '0' '1.6' '9.6' '2.4' '1.7' '1.1' '0.3' '1.5' '0.7' '1.2'
 '2.3' '0.5' '1.3' '0.2' '0.6' '1.4' '0.9' '1' '9.7']


user_score values include float, nan, and tbd. I will further explore the tbd value to see if it's associated with a specific year, country, or rating. 

In [63]:
# Check the dataset for patterns where user_score == tbd
display(vg[vg['user_score']=='tbd'].head(30))

name platform  year_of_release  \
119                              Zumba Fitness      Wii             2010   
301             Namco Museum: 50th Anniversary      PS2             2005   
520                            Zumba Fitness 2      Wii             2011   
645                               uDraw Studio      Wii             2010   
657   Frogger's Adventures: Temple of the Frog      GBA             <NA>   
718                            Just Dance Kids      Wii             2010   
726                  Dance Dance Revolution X2      PS2             2009   
821                            The Incredibles      GBA             2004   
881              Who wants to be a millionaire       PC             1999   
1047                             Tetris Worlds      GBA             2001   
1048                          Imagine: Teacher       DS             2008   
1069                    Personal Trainer: Math       DS             2007   
1070                              Game Party 3      Wii             2009   
1089                            Monsters, Inc.      GBA             2001   
1151         SpongeBob's Atlantis SquarePantis       DS             2007   
1199                            MySims Kingdom       DS             2008   
1210              Moshi Monsters: Moshling Zoo       DS             2011   
1284                  Disney Sing It: Pop Hits      Wii             2009   
1291                      Imagine: Master Chef       DS             2007   
1292                           Disney Princess      GBA             2003   
1298           The SpongeBob SquarePants Movie      GBA             2004   
1300                   Midway Arcade Treasures      PS2             2003   
1309                                    Wall-E       DS             2008   
1341           Disney Princess: Magical Jewels       DS             2007   
1367                         Hello Kitty Party       DS             2007   
1423                             SingStar Abba      PS2             2008   
1487               Petz Wild Animals: Dolphinz       DS             2007   
1541                 Imagine: Wedding Designer       DS             2008   
1560           Mystery Case Files: MillionHeir       DS             2008   
1700                            Disney Sing It      Wii             2008   

           genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  \
119       Sports      3.45      2.59      0.00         0.66           NaN   
301         Misc      2.08      1.35      0.00         0.54          61.0   
520       Sports      1.51      1.03      0.00         0.27           NaN   
645         Misc      1.65      0.57      0.00         0.20          71.0   
657    Adventure      2.15      0.18      0.00         0.07          73.0   
718         Misc      1.52      0.54      0.00         0.18           NaN   
726   Simulation      1.09      0.85      0.00         0.28           NaN   
821       Action      1.15      0.77      0.04         0.10          55.0   
881         Misc      1.94      0.00      0.00         0.00           NaN   
1047      Puzzle      1.25      0.39      0.00         0.06          65.0   
1048  Simulation      0.70      0.82      0.00         0.19           NaN   
1069      Puzzle      0.48      1.12      0.00         0.08          63.0   
1070      Puzzle      1.40      0.16      0.00         0.12          37.0   
1089   Adventure      0.59      0.95      0.03         0.10           NaN   
1151      Action      1.49      0.00      0.00         0.11          64.0   
1199  Simulation      0.80      0.59      0.01         0.16          58.0   
1210        Misc      0.37      0.96      0.00         0.21           NaN   
1284        Misc      1.07      0.28      0.00         0.13           NaN   
1291  Simulation      0.39      0.91      0.00         0.17           NaN   
1292    Platform      1.04      0.37      0.00         0.06           NaN   
1298    Platform      1.02      0.38      0.00         0.06         

Video Games with user_score of tbd are all associated with jp_sales of almost zero. I will check if there are any jp_sales of almost zero that have user ratings other than tbd to confirm whether this is the reason for the rating. 

In [64]:
# Check if all jp_sales close to zero have a user_score of tbd
display(vg[(vg['jp_sales'] >= 0) & (vg['jp_sales'] <= .05)].head(30))

name platform  year_of_release  \
60                Call of Duty: Ghosts     X360             2013   
61                        Just Dance 3      Wii             2011   
66                              Halo 4     X360             2012   
68                        Just Dance 2      Wii             2010   
72                           Minecraft     X360             2013   
78                              Halo 2       XB             2004   
85                          The Sims 3       PC             2009   
89                             Pac-Man     2600             1982   
99           Call of Duty: Black Ops 3     XOne             2015   
100         Call of Duty: World at War     X360             2008   
103                         Just Dance      Wii             2009   
116                       Just Dance 4      Wii             2012   
119                      Zumba Fitness      Wii             2010   
138                  World of Warcraft       PC             2004   
139                             Driver       PS             1999   
141                      Kinect Sports     X360             2010   
143                            FIFA 15      PS4             2014   
155  LEGO Star Wars: The Complete Saga      Wii             2007   
165                 Grand Theft Auto V     XOne             2014   
170         Call of Duty: World at War      PS3             2008   
176               Assassin's Creed III     X360             2012   
179     Call of Duty: Advanced Warfare     XOne             2014   
180                          Minecraft      PS3             2014   
183                    Madden NFL 2004      PS2             <NA>   
190                     FIFA Soccer 13     X360             2012   
192                         Diablo III       PC             2012   
195         Microsoft Flight Simulator       PC             1996   
196                     Guitar Hero II      PS2             2006   
197                          Fable III     X360             2010   
205             Tony Hawk's Pro Skater       PS             1999   

            genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  \
60        Shooter      6.73      2.56      0.04         0.91          73.0   
61           Misc      5.95      3.11      0.00         1.06          74.0   
66        Shooter      6.65      2.28      0.04         0.74          87.0   
68           Misc      5.80      2.85      0.01         0.78          74.0   
72           Misc      5.70      2.65      0.02         0.81           NaN   
78        Shooter      6.82      1.53      0.05         0.08          95.0   
85     Simulation      0.99      6.42      0.00         0.60          86.0   
89         Puzzle      7.28      0.45      0.00         0.08           NaN   
99        Shooter      4.59      2.11      0.01         0.68           NaN   
100       Shooter      4.81      1.88      0.00         0.69          84.0   
103          Misc      3.48      2.99      0.00         0.73          49.0   
116          Misc      4.05      2.16      0.00         0.55          74.0   
119        Sports      3.45      2.59      0.00         0.66           NaN   
138  Role-Playing      0.08      6.21      0.00         0.00          93.0   
139        Action      3.11      2.80      0.02         0.33          87.0   
141        Sports      3.92      1.73      0.03         0.51          73.0   
143        Sports      0.80      4.33      0.05         0.90          82.0   
155        Action      3.57      1.55      0.00         0.51          80.0   
165        Action      2.81      2.19      0.00         0.47          97.0   
170       Shooter      2.73      1.83      0.00         0.83          85.0   
176        Action      3.13      1.69      0.03         0.44          84.0   
179       Shooter      3.22      1.55      0.01         0.48          81.0   
180          Misc      2.03      2.37      0.00         0.87           NaN   
183        Sports      4.26      0.26      0.01      

There are user_scores other than tbd for other instances of jp_sales close to zero. It does not appear that these sales figures are the reason for the tbd rating. There do not appear to be any other patterns in year, genre, sales, critic_score, or rating that would explain the user_score of tbd. Because there are no clear patterns explaining this value, I will treat tbd as a missing value. 

In [65]:
# Fill user_score tbd with nan
vg['user_score'] = vg['user_score'].replace('tbd', np.nan)

print(vg['user_score'].unique())

['8' nan '8.3' '8.5' '6.6' '8.4' '8.6' '7.7' '6.3' '7.4' '8.2' '9' '7.9'
 '8.1' '8.7' '7.1' '3.4' '5.3' '4.8' '3.2' '8.9' '6.4' '7.8' '7.5' '2.6'
 '7.2' '9.2' '7' '7.3' '4.3' '7.6' '5.7' '5' '9.1' '6.5' '8.8' '6.9' '9.4'
 '6.8' '6.1' '6.7' '5.4' '4' '4.9' '4.5' '9.3' '6.2' '4.2' '6' '3.7' '4.1'
 '5.8' '5.6' '5.5' '4.4' '4.6' '5.9' '3.9' '3.1' '2.9' '5.2' '3.3' '4.7'
 '5.1' '3.5' '2.5' '1.9' '3' '2.7' '2.2' '2' '9.5' '2.1' '3.6' '2.8' '1.8'
 '3.8' '0' '1.6' '9.6' '2.4' '1.7' '1.1' '0.3' '1.5' '0.7' '1.2' '2.3'
 '0.5' '1.3' '0.2' '0.6' '1.4' '0.9' '1' '9.7']


In [66]:
# Change user_score to float type
vg['user_score'] = pd.to_numeric(vg['user_score'])

vg.dtypes

name                object
platform            object
year_of_release      Int64
genre               object
na_sales           float64
eu_sales           float64
jp_sales           float64
other_sales        float64
critic_score       float64
user_score         float64
rating              object
dtype: object

All variables are now the correct data type.

### 1c. Address Missing Values

In [67]:
# Count missing values in each variable
print('The number of missing values in each variable:')

print(vg.isna().sum())

The number of missing values in each variable:
name                  2
platform              0
year_of_release     269
genre                 2
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8578
user_score         9125
rating             6766
dtype: int64


There are missing values for name, year_of_release, genre, critic_score, user_score, and rating. 

In [68]:
# Display missing values for name
print('The missing values for name:')

display(vg[vg['name'].isna()])

The missing values for name:


name platform  year_of_release genre  na_sales  eu_sales  jp_sales  \
659    NaN      GEN             1993   NaN      1.78      0.53      0.00   
14244  NaN      GEN             1993   NaN      0.00      0.00      0.03   

       other_sales  critic_score  user_score rating  
659           0.08           NaN         NaN    NaN  
14244         0.00           NaN         NaN    NaN

The two rows missing on name are also the two rows missing on genre. These rows are additionally missing critic_score, user_score, and rating but do have complete information for platform, year, and sales. I will fill name and genre with 'unknown' and address the critic_score, user_score, and rating later on.

In [69]:
# Fill missing values in name with unknown
vg['name'] = vg['name'].fillna('unknown')

# Fill missing values in genre with unknown
vg['genre'] = vg['genre'].fillna('unknown')

In [70]:
# Display missing values for year_of_release
print('A sample of rows with missing values for year_of_release:')
display(vg[vg['year_of_release'].isna()].head(15))

A sample of rows with missing values for year_of_release:


name platform  year_of_release  \
183                               Madden NFL 2004      PS2             <NA>   
377                              FIFA Soccer 2004      PS2             <NA>   
456                    LEGO Batman: The Videogame      Wii             <NA>   
475                    wwe Smackdown vs. Raw 2006      PS2             <NA>   
609                                Space Invaders     2600             <NA>   
627                                     Rock Band     X360             <NA>   
657      Frogger's Adventures: Temple of the Frog      GBA             <NA>   
678   LEGO Indiana Jones: The Original Adventures      Wii             <NA>   
719                                Call of Duty 3      Wii             <NA>   
805                                     Rock Band      Wii             <NA>   
1131                      Call of Duty: Black Ops       PC             <NA>   
1142                                    Rock Band      PS3             <NA>   
1301                               Triple Play 99       PS             <NA>   
1506                                    Adventure     2600             <NA>   
1538                   LEGO Batman: The Videogame      PSP             <NA>   

          genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  \
183      Sports      4.26      0.26      0.01         0.71          94.0   
377      Sports      0.59      2.36      0.04         0.51          84.0   
456      Action      1.80      0.97      0.00         0.29          74.0   
475    Fighting      1.57      1.02      0.00         0.41           NaN   
609     Shooter      2.36      0.14      0.00         0.03           NaN   
627        Misc      1.93      0.33      0.00         0.21          92.0   
657   Adventure      2.15      0.18      0.00         0.07          73.0   
678      Action      1.51      0.61      0.00         0.21          78.0   
719     Shooter      1.17      0.84      0.00         0.23          69.0   
805        Misc      1.33      0.56      0.00         0.20          80.0   
1131    Shooter      0.58      0.81      0.00         0.23          81.0   
1142       Misc      0.99      0.41      0.00         0.22          92.0   
1301     Sports      0.81      0.55      0.00         0.10           NaN   
1506  Adventure      1.21      0.08      0.00         0.01           NaN   
1538     Action      0.57      0.44      0.00         0.27          73.0   

      user_score rating  
183          8.5      E  
377          6.4      E  
456          7.9   E10+  
475          NaN    NaN  
609          NaN    NaN  
627          8.2      T  
657          NaN      E  
678          6.6   E10+  
719          6.7      T  
805          6.3      T  
1131         5.2      M  
1142         8.4      T  
1301         NaN    NaN  
1506         NaN    NaN  
1538         7.4   E10+

There are not any apparent patterns in platform, genre, sales, score, or rating that explain the missing year_of_release values. A google search of the names shows that these games were released in all different years. Because a major portion of this analysis is to determine patterns based on year of release, filling these 236 values with the mean or median could skew the results in favor of that year. Therefore, I will leave these missing values and not use these games in analyses that include year_of_release.

In [71]:
# Display missing values for critic_score
print('A sample of missing values for critic_score:')

display(vg[vg['critic_score'].isna()].head(15))

A sample of missing values for critic_score:


name platform  year_of_release  \
1                         Super Mario Bros.      NES             1985   
4                  Pokemon Red/Pokemon Blue       GB             1996   
5                                    Tetris       GB             1989   
9                                 Duck Hunt      NES             1984   
10                               Nintendogs       DS             2005   
12              Pokemon Gold/Pokemon Silver       GB             1999   
18                        Super Mario World     SNES             1990   
20            Pokemon Diamond/Pokemon Pearl       DS             2006   
21                         Super Mario Land       GB             1989   
22                      Super Mario Bros. 3      NES             1988   
25            Pokemon Ruby/Pokemon Sapphire      GBA             2002   
27              Pokemon Black/Pokemon White       DS             2010   
30  Pokémon Yellow: Special Pikachu Edition       GB             1998   
31                Call of Duty: Black Ops 3      PS4             2015   
33                      Pokemon X/Pokemon Y      3DS             2013   

           genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  \
1       Platform     29.08      3.58      6.81         0.77           NaN   
4   Role-Playing     11.27      8.89     10.22         1.00           NaN   
5         Puzzle     23.20      2.26      4.22         0.58           NaN   
9        Shooter     26.93      0.63      0.28         0.47           NaN   
10    Simulation      9.05     10.95      1.93         2.74           NaN   
12  Role-Playing      9.00      6.18      7.20         0.71           NaN   
18      Platform     12.78      3.75      3.54         0.55           NaN   
20  Role-Playing      6.38      4.46      6.04         1.36           NaN   
21      Platform     10.83      2.71      4.18         0.42           NaN   
22      Platform      9.54      3.44      3.84         0.46           NaN   
25  Role-Playing      6.06      3.90      5.38         0.50           NaN   
27  Role-Playing      5.51      3.17      5.65         0.80           NaN   
30  Role-Playing      5.89      5.04      3.12         0.59           NaN   
31       Shooter      6.03      5.86      0.36         2.38           NaN   
33  Role-Playing      5.28      4.19      4.35         0.78           NaN   

    user_score rating  
1          NaN    NaN  
4          NaN    NaN  
5          NaN    NaN  
9          NaN    NaN  
10         NaN    NaN  
12         NaN    NaN  
18         NaN    NaN  
20         NaN    NaN  
21         NaN    NaN  
22         NaN    NaN  
25         NaN    NaN  
27         NaN    NaN  
30         NaN    NaN  
31         NaN    NaN  
33         NaN    NaN

All rows missing critic score in this sample are also missing user_score and rating. There appears to be a pattern in the missing values of these variables. Missing value imputation requires that there are not large patterns in missing values in the dataset. Therefore, imputing the appropriate mean, median, or mode for these values across the board would be a statistically inappropriate way to handle these values.

However, while these three variables had similar numbers of missing values, their counts weren't identical, meaning there are some rows that are missing only one or two of the scores. I will use imputation of mean or median for critic_score and user_score on rows that are missing only one or the other. For rows that are missing only user_score or critic_score, I will fill the rating value with 'unknown.'

Rows that are missing simultaneously on critic_score, user_score, and rating will be left out of analyses that use these values.

In [72]:
# Find the number of rows that are missing on critic_score, user_score, and rating
cs_us_rat_missing = len(vg[(vg['critic_score'].isna() & (vg['user_score'].isna()) & (vg['rating'].isna()))])

print(f'The number of rows missing on critic_score, user_score, and rating: {cs_us_rat_missing}')

The number of rows missing on critic_score, user_score, and rating: 6667


In [73]:
# Fill ratings in rows with either critic_score or user_score or both with 'unknown'
s1 = (vg['critic_score'].isna()) & (vg['user_score'].notna()) 
s2 = (vg['critic_score'].notna()) & (vg['user_score'].isna())
s3 = (vg['critic_score'].notna()) & (vg['user_score'].notna())

vg.loc[s1,'rating'] = vg.loc[s1, 'rating'].fillna(value='unknown') 
vg.loc[s2,'rating'] = vg.loc[s2, 'rating'].fillna(value='unknown') 
vg.loc[s3,'rating'] = vg.loc[s3, 'rating'].fillna(value='unknown')

print(vg.isna().sum())

name                  0
platform              0
year_of_release     269
genre                 0
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8578
user_score         9125
rating             6667
dtype: int64


In [160]:
# Check distribution for critic_score
cs_hist = px.histogram(vg, x='critic_score', title='Critic Score',
                        color_discrete_sequence=[px.colors.qualitative.D3[9]])

cs_hist.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

cs_hist.show()

critic_score is left-skewed, so I will fill missing values with the median.

In [109]:
cs_hist = px.histogram(vg, x='critic_score', color='genre', barmode='overlay',
                       title='Critic Score by Genre')

cs_hist.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

cs_hist.show()

In [76]:
cs_hist = px.histogram(vg, x='critic_score', color='platform', barmode='overlay',
                       title='Critic Score by Platform')

cs_hist.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

cs_hist.show()

There is a bigger difference in distribution of critic score by genre than by platform, so I will fill missing values of critic score with the median of the critic score by genre.

In [161]:
# Check distribution of user_score
us_hist = px.histogram(vg, x='user_score', title='User Score',
                       color_discrete_sequence=[px.colors.qualitative.D3[9]])

us_hist.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

us_hist.show()

user_score is highly left skewed, so I will fill missing values with the median.

In [149]:
cs_hist = px.histogram(vg, x='user_score', color='genre', barmode='overlay',
                       title='User Score by Genre')

cs_hist.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

cs_hist.show()

In [110]:
cs_hist = px.histogram(vg, x='user_score', color='platform', barmode='overlay',
                       title='User Score by Platform')

cs_hist.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

cs_hist.show()

The difference in distribution of user score is more distinct by platform than by genre, so I will fill the missing user scores with the median of user score by platform.

In [80]:
# Fill rows that are missing on user_score but not missing on critic_score with the median of user_score
s1 = vg['critic_score'].notna()

vg.loc[s1, 'user_score'] = vg.loc[s1, 'user_score'].fillna(vg.groupby('platform')['user_score'].transform('median'))

print(vg.isna().sum())

name                  0
platform              0
year_of_release     269
genre                 0
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8578
user_score         8005
rating             6667
dtype: int64


Missing user_scores are replaced with the median of user_score in rows where critic_score is present. User_scores have been left as missing values in rows that are also missing critic_score.

In [81]:
# Fill missing critic_scores that have a user_score with the median of critic_score
s1 = vg['user_score'].notna()

vg.loc[s1, 'critic_score'] = vg.loc[s1, 'critic_score'].fillna(vg.groupby('genre')['critic_score'].transform('median'))

In [82]:
# Print missing values that are left
print('After logical imputation, these are the missing values still in the dataset:')
print()
print(vg.isna().sum())

After logical imputation, these are the missing values still in the dataset:

name                  0
platform              0
year_of_release     269
genre                 0
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8005
user_score         8005
rating             6667
dtype: int64


After filling missing names and genres with 'unknown', filling critic_scores that had a user_score and user_scores that had a crtic_score with their respective medians, and 'unknown' for ratings that had either a critic_score or a user_score, these are the missing values left in the dataset. 

While there are still missing values, I argue that based on the analysis, it doesn't maek sense to impute the median for year. This is because year in this analysis is treated as categorical data, in that we will compare sales and scores across years. Therefore, imputing the mean or median of year would unduely weight that year's category. There are only 269 missing years. I will leave these values as missing.

Because of the large pattern in missing data for critic_score, user_score, and rating, I argue that it is not valid to impute values for games that are missing on all three values. I have imputed the median for critic_score where there was a user_score and the median user_score where there was a critic_score. I will assume that these missing values were due to another reason and not part of the larger pattern of missing across all three. 

Missing ratings were replaced with 'unknown' in rows that had values for at least one of critic_score or user_score.

I am confident in these missing value replacement strategies and will procede with the analysis using the remaining data.

### 1d. Check for Duplicates

In [187]:
# Check for implicit duplicate names
names = sorted(vg['name'].unique())

for name in names:
    print(name)

 beyblade burst
 fire emblem fates
 frozen: olaf's quest
 haikyu!! cross team match!
 tales of xillia 2
'98 koshien
.hack//g.u. vol.1//rebirth
.hack//g.u. vol.2//reminisce
.hack//g.u. vol.2//reminisce (jp sales)
.hack//g.u. vol.3//redemption
.hack//infection part 1
.hack//link
.hack//mutation part 2
.hack//outbreak part 3
.hack//quarantine part 4: the final chapter
.hack: sekai no mukou ni + versus
007 racing
007: quantum of solace
007: the world is not enough
007: tomorrow never dies
1 vs. 100
1/2 summer +
10 minute solution
100 all-time favorites
100 classic books
100 classic games
1000 cooking recipes from elle à table
1001 touch games
101-in-1 explosive megamix
101-in-1 party megamix wii
101-in-1 sports megamix
101-in-1 sports party megamix
1080°: teneighty snowboarding
11eyes: crossover
12-sai. honto no kimochi
12-sai. koisuru diary
12riven: the psi-climinal of integral
15 days
1701 a.d.
18 wheeler: american pro trucker
18 wheels of steel: extreme trucker
18 wheels of steel: extre

There do not appear to be any implicitly duplicated video game names.

In [83]:
# Check for fully duplicate rows
vg['name'] = vg['name'].str.lower()
vg['platform'] = vg['platform'].str.lower()

duplicates = vg.duplicated().sum()

print(f'The number of fully duplicate rows is: {duplicates}')

The number of fully duplicate rows is: 0


In [84]:
# Check for implicit duplicate name - platform - year duplicates
name_plat_duplicates = vg[['name', 'platform', 'year_of_release']].duplicated().sum()

print(f'The number of name-platform-year duplicates is: {name_plat_duplicates}')

The number of name-platform-year duplicates is: 2


In [85]:
# View the 2 duplicated rows
print('The two rows with duplicates are:')
display(vg[vg[['name', 'platform', 'year_of_release']].duplicated()==True])

The two rows with duplicates are:


name platform  year_of_release    genre  na_sales  eu_sales  \
14244        unknown      gen             1993  unknown       0.0      0.00   
16230  madden nfl 13      ps3             2012   Sports       0.0      0.01   

       jp_sales  other_sales  critic_score  user_score rating  
14244      0.03          0.0           NaN         NaN    NaN  
16230      0.00          0.0          83.0         5.5      E

In [86]:
# Display the original row and duplicate for the first duplicated row
print('The first duplicated rows are:')
display(vg[(vg['name']=='unknown') & (vg['platform']=='gen') & (vg['year_of_release']==1993)])

The first duplicated rows are:


name platform  year_of_release    genre  na_sales  eu_sales  \
659    unknown      gen             1993  unknown      1.78      0.53   
14244  unknown      gen             1993  unknown      0.00      0.00   

       jp_sales  other_sales  critic_score  user_score rating  
659        0.00         0.08           NaN         NaN    NaN  
14244      0.03         0.00           NaN         NaN    NaN

These two rows are identical other than na_sales, eu_sales, jp_sales, and other_sales. Because the sales figures are almost zero for the second row, I believe the second row is a mistake. I will delete the second row.

In [87]:
# Display the original row and the duplicate for the second duplicated row
print('The second duplicated rows are:')
display(vg[(vg['name']=='madden nfl 13') & (vg['platform']=='ps3') & (vg['year_of_release']==2012)])

The second duplicated rows are:


name platform  year_of_release   genre  na_sales  eu_sales  \
604    madden nfl 13      ps3             2012  Sports      2.11      0.22   
16230  madden nfl 13      ps3             2012  Sports      0.00      0.01   

       jp_sales  other_sales  critic_score  user_score rating  
604         0.0         0.23          83.0         5.5      E  
16230       0.0         0.00          83.0         5.5      E

These two rows are identical other than na_sales, eu_sales, jp_sales, and other_sales. Because the sales figures are almost zero for the second row, I believe the second row is a mistake. I will delete the second row.


In [88]:
# Drop the implicit duplicate rows
vg = vg.drop_duplicates(subset=['name', 'platform', 'year_of_release']).reset_index(drop=True)

In [89]:
# Check the duplicate name-platform-year are removed
name_plat_duplicates_2 = vg[['name', 'platform', 'year_of_release']].duplicated().sum()

print(f'The number of name-platform-year duplicates now is: {name_plat_duplicates_2}')

The number of name-platform-year duplicates now is: 0


### 1e. Add Additional Features

In [90]:
# Calculate total sum of sales across all regions
vg['total_sales'] = vg.iloc[:, 4:8].sum(axis=1)

vg.head()

name platform  year_of_release         genre  na_sales  \
0                wii sports      wii             2006        Sports     41.36   
1         super mario bros.      nes             1985      Platform     29.08   
2            mario kart wii      wii             2008        Racing     15.68   
3         wii sports resort      wii             2009        Sports     15.61   
4  pokemon red/pokemon blue       gb             1996  Role-Playing     11.27   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  \
0     28.96      3.77         8.45          76.0         8.0      E   
1      3.58      6.81         0.77           NaN         NaN    NaN   
2     12.76      3.79         3.29          82.0         8.3      E   
3     10.93      3.28         2.95          80.0         8.0      E   
4      8.89     10.22         1.00           NaN         NaN    NaN   

   total_sales  
0        82.54  
1        40.24  
2        35.52  
3        32.77  
4        31.38

## 2. Analyze the data

### 2a. Games Released by Year

In [91]:
# Calculate the number of games released by year
games_by_year = vg.groupby('year_of_release')['name'].count().reset_index()

display(games_by_year)

year_of_release  name
0              1980     9
1              1981    46
2              1982    36
3              1983    17
4              1984    14
5              1985    14
6              1986    21
7              1987    16
8              1988    15
9              1989    17
10             1990    16
11             1991    41
12             1992    43
13             1993    61
14             1994   121
15             1995   219
16             1996   263
17             1997   289
18             1998   379
19             1999   338
20             2000   350
21             2001   482
22             2002   829
23             2003   775
24             2004   762
25             2005   939
26             2006  1006
27             2007  1197
28             2008  1427
29             2009  1426
30             2010  1255
31             2011  1136
32             2012   652
33             2013   544
34             2014   581
35             2015   606
36             2016   502

In [162]:
# Create bar graph of video games released by year
year_bar = px.bar(games_by_year, x='year_of_release', y='name',
                  title='Games Released by Year', 
                  labels={'year_of_release':'Year of Release', 'name':'Count'},
                  color_discrete_sequence=[px.colors.qualitative.D3[9]],
                  width=800, height=500)

year_bar.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
}) 

year_bar.update_xaxes(showgrid=False)
year_bar.update_yaxes(range=[0, 1800], showgrid=False) 


year_bar.show()


This graph is highly left skewed, meaning not very many games were released each year from 1980 - 1995, while a majority of video games were released between 2000 and 2012. From the frequency table, we can see that fewer than 100 games were released per year before 1994. This makes sense, because video game consoles were not common household goods until the late 1990s. The number of games released per year increased from 1990 until 2008. The year with the highest number of video games released is 2008 with 1427 games released that year, closely followed by 2009 with 1426 game releases.  The number of games per year then saw a general decline from 2009 until the end of this data in 2016. 

### 2b. Total Sales by Platform

In [93]:
# Calculate total sales by platform
sales_by_plat = vg.groupby('platform')['total_sales'].sum().reset_index().sort_values('total_sales', ascending=False)

print(f'Total sales by platform from 1980 - 2016:')
display(sales_by_plat)

Total sales by platform from 1980 - 2016:


platform  total_sales
16      ps2      1255.77
28     x360       971.42
17      ps3       939.64
25      wii       907.51
4        ds       806.12
15       ps       730.86
6       gba       317.85
18      ps4       314.14
19      psp       294.05
13       pc       259.52
2       3ds       259.00
29       xb       257.74
5        gb       255.46
11      nes       251.05
10      n64       218.68
23     snes       200.04
7        gc       198.93
30     xone       159.32
0      2600        96.98
26     wiiu        82.19
20      psv        54.07
21      sat        33.59
8       gen        30.74
3        dc        15.95
22      scd         1.86
12       ng         1.44
27       ws         1.42
24     tg16         0.16
1       3do         0.10
9        gg         0.04
14     pcfx         0.03

In [163]:
# Create a bar chart for total sales by platform
plat_bar = px.bar(sales_by_plat, x='platform', y='total_sales', 
                  title='Total Sales by Platform',
                  labels={'total_sales':'Total Sales (USD million)', 'platform':'Platform'},
                  color_discrete_sequence=[px.colors.qualitative.D3[9]],
                  width=800, height=500)

plat_bar.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

plat_bar.update_layout(xaxis={'categoryorder':'total descending'}) # Arrange in order from expensive to inexpensive

plat_bar.update_xaxes(showgrid=False) # Turn off x grid
plat_bar.update_yaxes(range=[0, 1400], showgrid=False) # Turn off y grid

plat_bar.show()

From this graph, we can see a clear drop off betweeen the six highest selling platforms and all the rest. Those six highest-performing platforms are ps2, x360, ps3, wii, ds, and ps. I will use these six top selling platforms as my focus and further investigate sales per year by platform. 

In [95]:
# Create a data frame of top brands
top_brands = sales_by_plat.iloc[:6, 0]

vg_top_brand = vg.query('platform in @top_brands')

display(vg_top_brand['platform'].unique())

array(['wii', 'ds', 'x360', 'ps3', 'ps2', 'ps'], dtype=object)

In [96]:
# Create table for total sales by year by platform
sales_y_p = vg_top_brand.groupby(['year_of_release', 'platform'])['total_sales'].sum().reset_index()

print(sales_y_p.to_string())

    year_of_release platform  total_sales
0              1985       ds         0.02
1              1994       ps         6.03
2              1995       ps        35.96
3              1996       ps        94.70
4              1997       ps       136.17
5              1998       ps       169.49
6              1999       ps       144.53
7              2000       ps        96.37
8              2000      ps2        39.17
9              2001       ps        35.59
10             2001      ps2       166.43
11             2002       ps         6.67
12             2002      ps2       205.38
13             2003       ps         2.07
14             2003      ps2       184.31
15             2004       ds        17.27
16             2004      ps2       211.81
17             2005       ds       130.14
18             2005      ps2       160.66
19             2005     x360         8.25
20             2006       ds       119.81
21             2006      ps2       103.42
22             2006      ps3      

In [97]:
# Create a histogram of total sales by year by platform
sales_y_p_hist = px.histogram(vg_top_brand, x='year_of_release', y='total_sales',
                       color='platform', barmode='overlay',
                       title='Total Sales by Year by Platform',
                       labels={'total_sales':'Total Sales (US millions)', 'year_of_release':'Year of Release'},
                       width=1000, height=500)

sales_y_p_hist.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

sales_y_p_hist.update_traces(opacity=0.5)

sales_y_p_hist.update_xaxes(showgrid=False)
sales_y_p_hist.update_yaxes(range=[0, 250], showgrid=False)

sales_y_p_hist.show()

Based on the grouped table and graph, ps, ps2, and ds all peaked sales in the late 1990s and early 2000s and have zero sales in 2016. The manufactures likely discontinued these models in favor of the ps3 and 3ds. The distribution of sales by year by game shows that it generally takes about 10 years for a device to come on the market, peak, and then be discontinued.

### 2c. Determine which Years to Use

Based on the previous analysis, I argue it's appropriate to use the previous 5 years (2012-2016) to build a prognosis for 2017. The reasoning is two fold. First, the number of games released each year saw a sharp decline from 2011-2012. The number of games released per year has since stayed fairly consistent from 2012-2016. Second, devices tend to take about 10 years to come on the market, peak, and then delcline. Using more than 5 years time would catch the tail-end of popularity for devices that are now off the market. For example, the ds saw $145 USD millions sales in 2008, yet is now off the market. Additionally the wii saw $171 USD million sales in 2008 and has since dropped to almost $0 USD million in 2016. Starting the prediction period in 2012 misses the ds and wii boom yet catches the end of x360 and ps3 and will also catch any games that are now on the rise.

### 2d. Find Platforms that are Leading in Sales

In [98]:
# Subset data to include only the last 5 years
vg_5y = vg[vg['year_of_release'] >= 2012]

years_unique = sorted(vg_5y['year_of_release'].unique())

print(f'The data has been subset to include only these years: {years_unique}')

The data has been subset to include only these years: [2012, 2013, 2014, 2015, 2016]


In [99]:
# Calculate platform sales by year
sales_p_y_5 = vg_5y.groupby(['year_of_release', 'platform'])['total_sales'].sum().reset_index()

display(sales_p_y_5)

year_of_release platform  total_sales
0              2012      3ds        51.36
1              2012       ds        11.01
2              2012       pc        23.22
3              2012      ps3       107.35
4              2012      psp         7.69
5              2012      psv        16.19
6              2012      wii        21.71
7              2012     wiiu        17.56
8              2012     x360        99.74
9              2013      3ds        56.57
10             2013       ds         1.54
11             2013       pc        12.38
12             2013      ps3       113.25
13             2013      ps4        25.99
14             2013      psp         3.14
15             2013      psv        10.59
16             2013      wii         8.59
17             2013     wiiu        21.65
18             2013     x360        88.58
19             2013     xone        18.96
20             2014      3ds        43.76
21             2014       pc        13.28
22             2014      ps3        47.76
23             2014      ps4       100.00
24             2014      psp         0.24
25             2014      psv        11.90
26             2014      wii         3.75
27             2014     wiiu        22.03
28             2014     x360        34.74
29             2014     xone        54.07
30             2015      3ds        27.78
31             2015       pc         8.52
32             2015      ps3        16.82
33             2015      ps4       118.90
34             2015      psp         0.12
35             2015      psv         6.25
36             2015      wii         1.14
37             2015     wiiu        16.35
38             2015     x360        11.96
39             2015     xone        60.14
40             2016      3ds        15.14
41             2016       pc         5.25
42             2016      ps3         3.60
43             2016      ps4        69.25
44             2016      psv         4.25
45             2016      wii         0.18
46             2016     wiiu         4.60
47             2016     x360         1.52
48             2016     xone        26.15

In [100]:
# Create bar chart of sales by year of release by platform
sales_p_y_5_bar = px.histogram(vg_5y, x='year_of_release', y='total_sales', color='platform',
                               histfunc='sum', title='Total Sales by Platform by Year',
                               labels={'total_sales':'Total Sales (USD mills)', 'year_of_release':'Year of Release'},
                               width=1000, height=600)

sales_p_y_5_bar.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

sales_p_y_5_bar.show()

Total video game sales across all platforms genrally declined from 2012 through 2016. Ps3 and x360 had relatively large sales in 2012 and 2013 but have since dropped to almost $0 USD mills in 2016. 3ds had a slight decline from it's strongest sales in 2012 and 2013 but does still have notable sales in 2016 at $15.14 USD mills. However, both xone and ps4 were non-existant in 2012 and have seen substantial growth since their introduction in 2013. They both grew substantially from 2013 to 2014 and again slightly from 2014 to 2015. Both platforms took a slight decline in 2016, but the market as a whole took an even larger hit that year, and xone and ps4 still faired well in sales, bringing in $60.14 and $69.25 USD mils, respectively. Given that a platform's sales tend to start small and grow each year for about 5 years, I predict that xone and ps4 will continue to be sale leaders in 2017. I would additionally predict that 3ds will continue to 3rd spot with smaller, but next-largest, sales.

### 2e. Compare Global Total Sales by Platform

In [204]:
# Calculate Averge and Median Sales by Platorm with a Pivot Table
avg_platform = vg_5y.pivot_table(index='platform', values='total_sales', aggfunc=['mean', 'median']).round(2)

# Make pivot table into data frame
avg_platform.columns = avg_platform.columns.droplevel(1)

avg_platform.columns.name = None

avg_platform = avg_platform.reset_index().rename(columns={'mean':'mean_sales', 'median':'median_sales'})

display(avg_platform)

platform  mean_sales  median_sales
0       3ds        0.49          0.11
1        ds        0.40          0.05
2        pc        0.25          0.08
3       ps3        0.59          0.20
4       ps4        0.80          0.20
5       psp        0.06          0.03
6       psv        0.12          0.05
7       wii        0.65          0.18
8      wiiu        0.56          0.22
9      x360        0.81          0.31
10     xone        0.65          0.22

In [157]:
# Create bar chart comparing the mean and median sales for each platform
plat_mean_bar = px.bar(avg_platform, x='platform', y=['mean_sales', 'median_sales'], barmode='group',
                       title='Mean and Median Total Sales by Platform', 
                       labels={'value':'Total Sales (USD millions)', 'platform':'Platform'},
                       color_discrete_sequence=[px.colors.qualitative.D3[9],
                                                px.colors.qualitative.Plotly[1]],
                       width=1000, height=500)

plat_mean_bar.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

plat_mean_bar.update_layout(legend_title_text='Statistic', xaxis={'categoryorder':'total descending'})

plat_mean_bar.show()

In [152]:
# Print plotly colors to grab for charts
fig = px.colors.qualitative.swatches()
fig.show()

In [158]:
# Build a boxplot of sales for individual games by platform
sales_box = px.box(vg_5y, x='platform', y='total_sales', hover_data=['name'],
                   title='Global Sales by Platform',
                   labels={'total_sales':'Total Sales (USD million)', 'platform':'Platform'},
                color_discrete_sequence=[px.colors.qualitative.D3[9]],
                   width=1000, height=1300)

sales_box.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

sales_box.update_xaxes(showgrid=False)
sales_box.update_yaxes(range=[0, 23], showgrid=False)


The distribution of game sales by platform is highly right-skewed for all platforms. That means that for each platform, most of the games bring in $2 USD million or less, however there are a number of games that bring in very high sales. The platform with the highest upper outlier boundary is x360 at $2.05 USD million, meaning that is the Q3 + 1.5IQR cutoff for that platform. However, the upper outlier bounaries are not representative of the upper selling games for any of these platfoms. All platfoms have a large number of upper outliers. This  means that while a majority of their games make around $2 USD million or less, they all have a number of games that are high revenue leaders. 

For example, while the median total sales of games on ps3 is $0.20 USD million, meaning half of ps3 games bring in $0.20 USD million or less, ps3 also has Grand Theft Auto V that brings in a whopping $21.05 USD million. This is true for all of the platforms. As another example, wii has a median sales of $0.18 USD million, meaning half of wii games bring in $0.18 UST million or less, but it's highest gross game makes almost 40 times that amount at $7.09 USD million. 

Because of these extremely high sellers, the platforms are able to bring their average revenue up far above their median revenue. The extreme high earners pull the average up. This can be seen in the bar chart diplaying median vs average revenue for each platform. The average revenues are about $0.40 USD million higher than the median revenues for each platform. The high skew means that the median is a better statistical measure of center for each platform, but if the platforms wanted to report their sales figures in reports or to stakeholders, they would look better presenting the mean. 

Comparing the platforms, x360 and ps4 have the highest mean sales, while x360, xone, and wiiu have the highest median sales. When looking at both the mean and the median, there are 6 clear top performers: x60, xone, ps3, ps4, wii, and wiiu. However, when looking at the boxplots, ps3, x360, ps4, and 3ds have the highest grossing individual games by far. 

### 2f. User and Critic Reviews vs Total Sales

In [164]:
# Scatter plot of total_sales vs user_score for xone
vg_5y_xone = vg_5y[vg_5y['platform']=='xone']

user_scat = px.scatter(vg_5y_xone, x='user_score', y='total_sales', hover_data=['name'],
                       title='Total Sales vs User Score for Xone',
                       labels={'total_sales':'Total Sales (USD million)', 'user_score':'User Score'},
                       color_discrete_sequence=[px.colors.qualitative.D3[9]],
                       width=1000, height=600)

user_scat.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

user_scat.update_xaxes(showgrid=False)
user_scat.update_yaxes(showgrid=False)

user_scat.show()

In [165]:
# Scatter plot of total_sales vs critic_score for xone
crit_scat = px.scatter(vg_5y_xone, x='critic_score', y='total_sales', hover_data=['name'],
                       title='Total Sales vs Critic Score for Xone',
                       labels={'total_sales':'Total Sales (USD million)', 'critic_score':'Critic Score'},
                       color_discrete_sequence=[px.colors.qualitative.D3[9]],
                       width=1000, height=600)

crit_scat.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

crit_scat.update_xaxes(showgrid=False)
crit_scat.update_yaxes(showgrid=False)

crit_scat.show()

In [108]:
# Calculate correlation coefficients
vg_5y[['total_sales','user_score', 'critic_score']].corr()

total_sales  user_score  critic_score
total_sales      1.000000    0.003000      0.307612
user_score       0.003000    1.000000      0.430633
critic_score     0.307612    0.430633      1.000000

The relationshp between user_score, critic_score, and total_sales can be seen in both the scatter plots and the correlation coefficients. Based on the scatter plots, user_score does not appear to be strongly related to total_sales. No matter the user_scores, there are games that make both really low and really high total_sales. This is also reflected in the correlation coeffient of 0.0030, which repersents an extremely weak, positive relationship between user_score and total_sales.

The relationship is slighly stronger between critic_score and total_sales. From the scatter plot, we can see there is a generally positive relationship between critic_scores and sales, meaning that games that receive higher critic scores also tend to bring in higher total sales. This is also refelcted in the correlation coefficient of 0.3076, which represents a weak, positive relationship between critic_score and total_sales.

Based on these plots and correlations, neither user_score nor critic_score are particularly strong predictors of total_sales. However, comparing the two, critic_score is a better predictor of total_sales than user_score. This means if someone wanted to predict how much revenue a game will bring in, the score that game got from the critics would be a better measure of how many total sales the game will have than the user_score.

### 2g. Compare Sales of Games on Different Platforms

In [129]:
# Find the 10 highest-selling games
game_best_sellers = vg_5y.groupby('name')['total_sales'].sum().reset_index().sort_values('total_sales', ascending=False)

# Print the top 10 grossing games and their total sales
display(game_best_sellers.head(10))

name  total_sales
558              grand theft auto v        56.58
183      call of duty: black ops ii        29.40
184            call of duty: ghosts        27.39
182       call of duty: black ops 3        25.67
873                       minecraft        24.16
181  call of duty: advanced warfare        21.97
467                         fifa 15        17.37
466                         fifa 14        16.46
468                         fifa 16        16.30
471                  fifa soccer 13        15.97

In [172]:
# Create list of top 10 games
games_top10 = game_best_sellers.iloc[:10, 0]

# Create a dataframe of top 10 games, platform, and total_sales
vg_5y_top10 = vg.query('name in @games_top10')

print('Tope 10 games, their platforms, and total_sales on each platform:')
print(vg_5y_top10[['name', 'platform', 'total_sales']].sort_values(['name', 'total_sales'], ascending=False).to_string())

Tope 10 games, their platforms, and total_sales on each platform:
                                name platform  total_sales
72                         minecraft     x360         9.18
180                        minecraft      ps3         5.27
261                        minecraft      ps4         4.32
543                        minecraft     xone         2.76
868                        minecraft      psv         1.96
2973                       minecraft     wiiu         0.67
16                grand theft auto v      ps3        21.05
23                grand theft auto v     x360        16.27
42                grand theft auto v      ps4        12.62
165               grand theft auto v     xone         5.47
1730              grand theft auto v       pc         1.17
81                    fifa soccer 13      ps3         8.17
190                   fifa soccer 13     x360         5.16
2945                  fifa soccer 13      wii         0.69
4153                  fifa soccer 13     wiiu    

In [151]:
# Create bar chart comparing total_revenue per game by platform
games_bar = px.bar(vg_5y_top10, x='name', y='total_sales', color='platform',
                         barmode='group', title='Top 10 Games Total Sales by Platform',
                         labels={'total_sales':'Total Sales (USD millions)', 'name':'Video Game'},
                         height=700)

games_bar.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

games_bar.update_xaxes(showgrid=False)
games_bar.update_yaxes(showgrid=False)

games_bar.show()

From the total sales by game data frame and bar chart, we can see that diffent games do better on different platforms. For example, grand theft auto v, call of duty: black ops ii, call of duty: ghosts, fifa soccer 13, fifa 14, and minecraft all made the most revenue on games sold for ps3 and x360. Call of duty: advanced warfare, fifa 16, and call of duty: black ops all brought in the most revenue on games for the ps4. While ps3, x360, ps4, and xone are generally the highest revenue platforms, which platform brings in the most sales does depend on the game itself. It's possible that the year of release of a game is part of this relationship as well. Games released in later years are likely to make more sales on the newest versions of the platforms, like ps4, and games that were released in earlier years are likely to make more sales on the older platforms, like ps3.

### 2h. Total Sales by Genre

In [145]:
# Calculate mean and median of sales by genre
avg_genre = vg_5y.pivot_table(index='genre', values='total_sales', aggfunc=['mean', 'median']).round(2)

# Make pivot table into data frame
avg_genre.columns = avg_genre.columns.droplevel(1)

avg_genre.columns.name = None

avg_genre = avg_genre.reset_index().rename(columns={'mean':'mean_sales', 'median':'median_sales'})

display(avg_genre)

genre  mean_sales  median_sales
0         Action        0.43          0.12
1      Adventure        0.10          0.03
2       Fighting        0.41          0.13
3           Misc        0.44          0.12
4       Platform        0.72          0.21
5         Puzzle        0.17          0.04
6         Racing        0.47          0.14
7   Role-Playing        0.52          0.14
8        Shooter        1.30          0.44
9     Simulation        0.44          0.12
10        Sports        0.68          0.24
11      Strategy        0.19          0.08

In [168]:
# Create bar chart of mean and median of sales by genre
genre_mean_bar = px.bar(avg_genre, x='genre', y=['mean_sales', 'median_sales'], barmode='group',
                       title='Mean and Median Total Sales by Genre', 
                       labels={'value':'Total Sales (USD millions)', 'genre':'Genre'},
                       color_discrete_sequence=[px.colors.qualitative.D3[9],
                                                px.colors.qualitative.Plotly[1]],
                       width=1000, height=500)

genre_mean_bar.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

genre_mean_bar.update_layout(legend_title_text='Statistic', xaxis={'categoryorder':'total descending'})

genre_mean_bar.show()

In [192]:
# Create a histogram of sales by genre
sales_genre_box = px.box(vg_5y, x='genre', y='total_sales', hover_data=['name'],
                         title='Total Sales by Genre',
                         labels={'total_sales':'Total Sales (USD millions)', 'genre':'Genre'},
                        color_discrete_sequence=[px.colors.qualitative.D3[9]],
                        width=1000, height=1200)

sales_genre_box.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

sales_genre_box.update_xaxes(showgrid=False)
sales_genre_box.update_yaxes(range=[0, 25], showgrid=False)

sales_genre_box.show()

From the mean and median total sales by genre bar graph, we can see that the genres with the 3 highest mean and median revenue per video game are shooter, platform, and sports.  However, when looking at the boxpolots, we can see that the genre with the single highest-performing game is action with grand theft auto v bringing in $21.05 USD million. In general, however, action tends to under perform compared to shooter, platform, and sports genres. The median, third quartile, and upper-outlier boundary for action all fall below these same values for shooter, platform, and sports. If someone were only concerned about the highest revenue game in a genre, then action, shooter, and role-playing would come out on top. However, if they are concerned about the general distribution and center of total sales, then shooter, platform, and sports come out on top. 

The genres with the lowest 3 mean and median revenue per video game are strategy, puzzle, and adventure. No matter whether we look at center or distribution of total sales by game, strategy, adventure, and puzzle come out on the bottom. They have both the lowest means and medians of any genre and also the lowest top-selling games topping out at around $1.5 USD million for the highest selling game across all three genres.

In general, businesses can expect to get the highest sales from Action, Shooter, Role-Playing, Platform, and Sports genres and the lowest sales from Strategy, Advendure, and Puzzle games.

## 3. Create a User Profile for Each Region

### 3.1 Top 5 Platforms by Region

In [230]:
# Find total sales of each platform in each region
total_sales_region = vg_5y.pivot_table(index='platform', values=['na_sales', 'eu_sales', 'jp_sales'], aggfunc='sum')

total_sales_region = total_sales_region.transpose().reset_index().rename(columns={'platform':'index', 'index':'region'})

display(total_sales_region)

platform    region    3ds    ds     pc     ps3     ps4    psp    psv    wii  \
0         eu_sales  42.64  3.53  37.76  106.85  141.09   0.42  11.36  11.92   
1         jp_sales  87.79  3.72   0.00   35.29   15.96  10.47  21.04   3.39   
2         na_sales  55.31  4.59  19.12  103.38  108.74   0.13  10.98  17.45   

platform   wiiu    x360   xone  
0         25.13   74.52  51.59  
1         13.01    1.57   0.34  
2         38.10  140.05  93.12

In [238]:
# Create bar chart of total platform sales by region
sales_reg_bar = px.bar(total_sales_region, x='region', y=['x360', 'ps4', 'ps3', 'xone', '3ds', 'wiiu', 'pc', 'ps3', 'psv', 'wii', 'ds', 'psp',],
                       barmode='group')

sales_reg_bar.update_layout({
    'plot_bgcolor':'rgba(0, 0, 0, 0)',
    'paper_bgcolor':'rgba(0, 0, 0, 0)'
})

sales_reg_bar.update_layout(legend_title_text='Region', xaxis={'categoryorder':'total descending'})

sales_reg_bar.update_xaxes(showgrid=False)
sales_reg_bar.update_yaxes(showgrid=False)

sales_reg_bar.show()

In [ ]:
# List the top 5 platforms by region 

## 4. Test Hypotheses

## 5. Conclusion